File: typescript/rotational-cipher/rotational-cipher.ts
const lowerCaseLetters = 'abcdefghijklmnopqrstuvwxyz'
const upperCaseLetters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

export default class RotationalCipher {
    static rotate(plaintext: string, rotation: number) {
        return Array.prototype.map.call(plaintext, (letter: string) => this.rotateChar(letter, rotation)).join('')
    }

    static rotateChar(char: string, rotation: number) {
        if (!this.isLetter(char))
            return char;

        const key = char.toLowerCase() === char ? lowerCaseLetters : upperCaseLetters
        return key[(key.indexOf(char) + rotation) % key.length]
    }

    static isLetter(char: string) {
        return char.toLowerCase() !== char.toUpperCase()
    }
}

File: ruby/rotational-cipher/rotational_cipher.rb
module RotationalCipher
  def self.rotate(plaintext, by)
    rotate_char = shift_map([*'a'..'z'], by).merge(shift_map([*'A'..'Z'], by))
    plaintext.chars.map {|c| rotate_char.fetch(c, c)}.join
  end

  private
  def self.shift_map(array, by) = array.zip(array.rotate(by)).to_h
end